integreat_compass/cms/views/index.py

```diff
@@ -1,9 +1,93 @@
+from ..constants import offer_group_types, offer_mode_types
+from ..models import Offer, Tag
+
+
+class IndexListView(ListView):
     """
-    Start page of the project
+    View to see all offers
     """
 
     template_name = "index.html"
+    context_object_name = "all_offers_list"
+
+    def get_queryset(self):
+        """
+        :return: The offer queryset
+        :rtype: ~django.db.models.query.QuerySet
+        """
+        return Offer.objects.all()
+
+    def get_context_data(self, **kwargs):
+        """
+        This functions displays all offers
+        :param kwargs: The supplied keyword arguments
+        :type kwargs: dict
+
+        :return: The context of all
+        :rtype: dict
+        """
+        offers = Offer.objects.all()
+        tags = Tag.objects.all()
+        print(type(Offer.objects.all()))
+        context = {
+            "tags": tags,
+            "offers": offers,
+            "offer_group_types": offer_group_types.CHOICES,
+            "offer_mode_types": offer_mode_types.CHOICES,
+        }
+        return context
+
+    def get(self, request, *args, **kwargs):
+        """
+        This functions filters
+
+        :param request: The current request
+        :type request: ~django.http.HttpRequest
+
+        :param args: The supplied arguments
+        :type args: list
+
+        :param kwargs: The supplied keyword arguments
+        :type kwargs: dict
+
+        :return: Response for filtered offers
+        :rtype: ~django.template.response.TemplateResponse
+        """
+        tags = Tag.objects.all()
+        filtered_offers = Offer.objects.all()
+        if request.GET.getlist("offer-tags"):
+            filtered_offers = reduce(
+                lambda offers, tag: filtered_offers.filter(tags=tag),
+                request.GET.getlist("offer-tags"),
+                filtered_offers,
+            )
+        if request.GET.get("mode-type"):
+            filtered_offers = Offer.objects.filter(
+                mode_type__in=request.GET.getlist("mode-type")
+            ).distinct()
+        if request.GET.get("group-type"):
+            filtered_offers = Offer.objects.filter(
+                group_type__in=request.GET.getlist("group-type")
+            ).distinct()
+        if request.GET.get("offer-search"):
+            filtered_offers = Offer.objects.filter(
+                versions__title__icontains=request.GET.get("offer-search")
+            ).distinct()
+        return render(
+            request,
+            self.template_name,
+            {
+                "tags": tags,
+                "offers": filtered_offers,
+                "filtered_tags": request.GET.getlist("offer-tags"),
+                "offer_group_types": offer_group_types.CHOICES,
+                "filtered_group_types": request.GET.getlist("group-type"),
+                "offer_mode_types": offer_mode_types.CHOICES,
+                "filtered_mode_types": request.GET.getlist("mode-type"),
+            },
+        )
```

Comment:
```suggestion
        filtered_offers = Offer.objects.all()
        if offer_tags := request.GET.getlist("offer-tags"):
            tag_filters = Q(tags__in=offer_tags)
            filtered_offers = filtered_offers.filter(tag_filters)

        if mode_types := request.GET.getlist("mode-type"):
            mode_types_filter = Q(mode_type__in=mode_types)
            filtered_offers = filtered_offers.filter(mode_types_filter)

        if group_types := request.GET.getlist("group-type"):
            group_types_filter = Q(group_type__in=group_types)
            filtered_offers = filtered_offers.filter(group_types_filter)

        if offer_search := request.GET.get("offer-search"):
            title_contains_filter = Q(versions__title__icontains=offer_search)
            filtered_offers = filtered_offers.filter(title_contains_filter)

        return render(
            request,
            self.template_name,
            {
                "tags": Tag.objects.all(),
                "offers": filtered_offers.distinct(),
                "filtered_tags": offer_tags,
                "offer_group_types": offer_group_types.CHOICES,
                "filtered_group_types": group_types,
                "offer_mode_types": offer_mode_types.CHOICES,
                "filtered_mode_types": mode_types,
            },
        )
```

This is a suggestion for making the code a little more readable. At the same time, I am having an issue with the (current!) implementation.

For example, click "Berufssprache" -> filter works correctly.
Then click "Gesellschaft" -> only "Deutsch für Anfänger" is shown.

This makes no sense, since either both should be shown (offers matching at least one tag), or none (offers matching all tags).

I do not really know what the reason is.

The suggested code implements the "matching at least one tag" tag. I could not get the "matching all tags" to work, unsure what the reason is.